### ISSUE TITLE
Virtualize RepositoryBase methods

### ISSUE THREAD

**Issue by ngruson** [ardalis/Specification#66]
What do you feel about making the methods in RepositoryBase.cs virtual?
I ran into a NullReferenceException when unit testing my concrete EF6 implementation of the repository with a mocked DbContext.
```csharp
public virtual async Task UpdateAsync(T entity)
{
  dbContext.Entry(entity).State = EntityState.Modified; // Entry call to mocked DbContext returns null

  await SaveChangesAsync();
}
```
I'm not able to mock the Entry method of DbContext because the constructor of DbEntityEntry is not public.

If UpdateAsync was virtual in RepositoryBase, I could override it in my concrete repository:
```csharp
public override async Task UpdateAsync(T entity)
{
  ((MyDbContext)DbContext).SetModified(entity);
  await SaveChangesAsync();
}
```
SetModified in my concrete repository:
```csharp
public virtual void SetModified(object entity)
{
  Entry(entity).State = EntityState.Modified;
}
```
I would need a protected property to get to the DbContext.
In this setup, I'm able to unit test my concrete repository without issues.

**Comment by fiseni** [comment]
Hi @ngruson,

It's not about the method implementations only, but also the namespaces used within the "RepositoryBase".

The `EntityFramework6` project should fully implement `IRepositoryBase` interface, and not inherit from the existing one in `EntityFrameworkCore` project. These two "addon/plugin" projects shouldn't reference each other. As for making the methods `virtual`, sure, why not.

Also, we should have a new project for tests, something like `EntityFramework6.IntegrationTests`. That way we can clearly test upon EF6, without referencing the EFCore at all.

This is my stand initially, @ardalis may have a different opinion.

PS. We have to refactor the test infrastructure here. It's in our backlog for some time. We should have the following:
- UnitTests project for the base package.
- IntegrationsTests project for each plugin package.

**Comment by ngruson** [comment]
Hi @fiseni 

Just to clarify further:
The EF6 base repository is not referencing the EF Core base repository and is implementing IRepositoryBase.
In my solution I have a repository which is inheriting from RepositoryBase (it's EF6 but it might as well be EF Core).
I have unit tests in my solution to test this repository. So this is not about the tests in Ardalis.Specification.

Shall I make a PR to virtualize these methods in both the EF Core and EF6 RepositoryBase classes?

**Comment by fiseni** [comment]
Lets' first wait to merge the existing PR, and fix the EF6 project properly. Once we sort that out, then you can create an additional PR for these changes. Sounds OK?

For now, please use your own fully implemented Repository (you may just copy the implementations and update them as you wish). Anyhow, the `RepositoryBase` is given just as an optional base class and just as a reference. We try to avoid leading or enforcing users to use it.

**Comment by fiseni** [comment]
The methods of base repositories now are marked as virtual in `v5`.

**Comment by fiseni** [comment]
Implemented in `version 5.0.3`
